### PR TITLE
usb: store alternate interface settings

### LINKED ISSUE #24200: USB GET_INTERFACE response always 0, even when an alternate setting is used
**Describe the bug**

The response to a USB GET_INTERFACE is always 0, see the following excerpt from subsys/usb/usb_device.c. 

```
	case REQ_GET_INTERFACE:
		/* there is only one interface, return n-1 (= 0) */
		data[0] = 0U;
		*len = 1;
		break;
```
But according to USB Specification Revision 2.0 (usb_20.pdf from usb.org) section 9.4.4 GET_INTERFACE should return "selected alternate setting for the specified interface."

**To Reproduce**

There is no (or at least I can't find one that builds for my board) sample that has multiple USB alternate settings; but it is possible to reproduce using a patch to subsys/usb/class/loopback.c and samples/subsys/usb/testusb. 

1. Apply the following patch.

```
diff --git a/subsys/usb/class/loopback.c b/subsys/usb/class/loopback.c
index 754c2a4906..df35aee390 100644
--- a/subsys/usb/class/loopback.c
+++ b/subsys/usb/class/loopback.c
@@ -30,6 +30,7 @@ struct usb_loopback_config {
        struct usb_if_descriptor if0;
        struct usb_ep_descriptor if0_out_ep;
        struct usb_ep_descriptor if0_in_ep;
+       struct usb_if_descriptor if0_1;
 } __packed;

 USBD_CLASS_DESCR_DEFINE(primary, 0) struct usb_loopback_config loopback_cfg = {
@@ -65,6 +66,19 @@ USBD_CLASS_DESCR_DEFINE(primary, 0) struct usb_loopback_config loopback_cfg = {
                .wMaxPacketSize = sys_cpu_to_le16(CONFIG_LOOPBACK_BULK_EP_MPS),
                .bInterval = 0x00,
        },
+
+       .if0_1 = {
+               .bLength = sizeof(struct usb_if_descriptor),
+               .bDescriptorType = USB_INTERFACE_DESC,
+               .bInterfaceNumber = 0,
+               .bAlternateSetting = 1,
+               .bNumEndpoints = 0,
+               .bInterfaceClass = CUSTOM_CLASS,
+               .bInterfaceSubClass = 0,
+               .bInterfaceProtocol = 0,
+               .iInterface = 0,
+       },
+
 };
 /* usb.rst config structure end */
```
2. Build the testusb sample (my example below for the NXP RT1020 EVK). 

```
west build -p auto -b mimxrt1020_evk samples/subsys/usb/testusb
```
3. Flash.
4. Run USB Chapter 9 TEsts (USB 2 Devices) in the compliance tool from usb.org (USB 3 Gen X).

The test will fail in Configured State - TD 9.4 with an error message "A successful GetInterface request must return the alternate setting set by a prior call to SetInterface". 

**Expected behavior**

Compliance tests should pass.

**Environment (please complete the following information):**
 - OS: Ubuntu
 - Toolchain: Zephyr SDK 0.11.2
 - Zephyr master b74cad5a5d0f35905c4ab3d719bdc6b2794bc940.

### PR DESCRIPTION
**Pull request by jfischer-no**
Fixes request hijack through Audio class workaround in composite configuration.
Fixes: #24200

### CODE REVIEW

**Review by eobalski — CHANGES_REQUESTED** [review]
Part of the API documentation must to be updated, especially [this](https://github.com/zephyrproject-rtos/zephyr/blob/main/include/usb/usb_device.h#L158L166).

**Comment by jfischer-no** [comment]
> Part of the API documentation must to be updated, especially [this](https://github.com/zephyrproject-rtos/zephyr/blob/main/include/usb/usb_device.h#L158L166).

No, there is no reason to update.

File: subsys/usb/usb_device.c [hunk 1]
```diff
@@ -1389,12 +1437,22 @@ static int custom_handler(struct usb_setup_packet *pSetup,
...
 		 */
-		if (iface->custom_handler &&
-		    (if_descr->bInterfaceNumber == (pSetup->wIndex & 0xFF) ||
-		     if_descr->bInterfaceClass == AUDIO_CLASS)) {
+		if (if_descr->bInterfaceClass == AUDIO_CLASS) {
+			(void)iface->custom_handler(pSetup, len, data);
+		}
+
+		if (if_descr->bInterfaceNumber == (pSetup->wIndex & 0xFF)) {
 			return iface->custom_handler(pSetup, len, data);
```
Comment (anangl) [inline subsys/usb/usb_device.c:1456]: Is this intentional that now the handler will be called twice for this audio interface?

Comment (jfischer-no) [reply subsys/usb/usb_device.c:1456]: I think it can be moved to else branch now, I will change it.

File: subsys/usb/class/audio/audio.c [hunk 2]
```diff
@@ -689,6 +689,11 @@ static int audio_custom_handler(struct usb_setup_packet *pSetup, int32_t *len,
 
 	uint8_t iface = (pSetup->wIndex) & 0xFF;
 
+	if (REQTYPE_GET_RECIP(pSetup->bmRequestType) ==
+	    REQTYPE_RECIP_INTERFACE) {
+		return -EINVAL;
+	}
+
```
Comment (emob-nordic) [inline subsys/usb/class/audio/audio.c:696]: Wouldn't all interface request be dropped by `custom_handler` right now? Wouldn't the class be unable to set `rx/tx_enable` flags by this addition?

Comment (jfischer-no) [reply subsys/usb/class/audio/audio.c:696]: Yes, thanks, that should be inverse, also `!= REQTYPE_RECIP_INTERFACE`.

File: subsys/usb/usb_device.c [hunk 3]
```diff
@@ -1437,13 +1437,23 @@ static int custom_handler(struct usb_setup_packet *pSetup,
...
-		if (iface->custom_handler &&
-		    (if_descr->bInterfaceNumber == (pSetup->wIndex & 0xFF) ||
-		     if_descr->bInterfaceClass == AUDIO_CLASS)) {
+		if (iface->custom_handler == NULL) {
+			continue;
+		}
+
+		if (if_descr->bInterfaceNumber == (pSetup->wIndex & 0xFF)) {
 			return iface->custom_handler(pSetup, len, data);
+		} else {
```
Comment (emob-nordic) [inline subsys/usb/usb_device.c:1446]: you can concatenate this to single `else if` instead. No need to add braces here imho.

**Review by eobalski — APPROVED** [review]
Should work, however i can't say i like that. It should be done the other way around. If there was an API for writing to interface instead of endpoint directly, the class could then receive something like not enabled interface or similar error code. Having this audio class would be simplified by not having those 2 flags with `tx/rx_enable`. This is just a thinking loud comment. I see this PR as a workaround and maybe someday there would be a time and need to do that.

**Review by anangl — APPROVED** [review]

**Pull request merged**